File: question/neetcode150/sliding_window/sliding-window-maximum.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& a, int k) {
        deque<pair<int,int>> q;
        int n = a.size();

        vector<int> ans;

        for(int r=0;r<n;r++){
            int x = a[r];
            while(!q.empty() and q.front().first <= r-k){
                q.pop_front();
            }
            while(!q.empty() and q.back().second <= x){
                q.pop_back();
            }

            q.push_back({r,x});
            if (r+1>=k)
                ans.push_back(q.front().second);
        }

        return ans;
    }
};

File: question/neetcode150/sliding_window/sliding-window-maximum.py
from typing import List
from collections import deque

class Solution:
    def maxSlidingWindow(self, a: List[int], k: int) -> List[int]:
        q = deque()
        n = len(a)
        ans = []
        for r,n in enumerate(a):
            l = r-k
            while q and q[0][0] <= l:
                q.popleft() 
            while q and q[-1][1] < n:
                q.pop()
            q.append((r,n))

            if r+1 >= k:
                ans.append(q[0][1])
        
        return ans